### DISCUSSION TITLE
CI/CD processes for python libraries

### DISCUSSION THREAD

**Discussion by thclark** [octue/.github#14] [Python Libraries and Applications]
This discussion is an ongoing thread about the latest developments in our CI/CD processes. 

The aim is that when we take over maintenance of a library, or we come back to maintaining something we haven't touched for a while, we can skim down the notes here to catch up on what updates to our best practices have been made since the last time we touched that library, including pointers to where in our libraries you'd find and example of the latest best practice. 

**SORT BY NEWEST FIRST TO SEE MOST RECENT UPDATES!**

**Comment by thclark** [comment]
## Trusted publishing

@cortadocodes updated [django-twined](https://github.com/octue/django-twined/pull/78) to use trusted publishing, featuring:

- Publish to test pypi is done in the `ci` workflow because you want to know the wheel builds before you merge the code.
- Publish to main pypi is done in the `cd` workflow. It's done *after* the release because:
  - it's not possible to "atomically" to two steps at once, but
  - it *is* possible to manually revert the release in the event that the publish step fails for some reason.

**Comment by thclark** [comment]
## CodeQL

@cortadocodes updated [django-twined (in this pr)](https://github.com/octue/django-twined/pull/78) to add CodeQL analysis.

This isn't considered a 'check' on a PR, but rather gives a general status badge for the repository. It's in a separate `codeql` workflow for that reason, triggered on push to `main` so it simply shows the latest status of the repository.

**Comment by cortadocodes** [comment]
## Semantic version checker

@cortadocodes added the `check-semantic-version` reusable workflow as an independent job in the `ci` workflow for `django-twined` [in this PR](https://github.com/octue/django-twined/pull/78).

Compared to putting it in its own `semantic` workflow, this simplifies the GitHub check text from `semantic/check-semantic-version/check-semantic-version` to `ci/check-semantic-version/check-semantic-version`, which is a bit more readable and less repetitive.

**Reply by thclark** [reply]
@cortadocodes put an L2 header in these notes (two hashes) so we can see at a glance what they're about. I'll edit this one

**Comment by thclark** [comment]
## Running pre-commit in CI

We should be runnign precommit in CI. Here's why and how:
https://github.com/octue/django-gcp/pull/71#discussion_r1822685981

**Comment by thclark** [comment]
## Test Publishing

For the publish-test actions, we should be testing in parallel with any actual test matrix. First because it tests a different thing, Second because it gives us a clearer view of the complete picture. Discussion here:
https://github.com/octue/django-gcp/pull/71#discussion_r1822709259

**Comment by cortadocodes** [comment]
## Import sorting via `ruff` pre-commit
I noticed that the import sorting (isort) part of the `ruff` pre-commit checks wasn't happening (possibly only when not running in a devcontainer). The fix is to update the `ruff` section in `.pre-commit-config.yaml` like this:

```
  - repo: https://github.com/astral-sh/ruff-pre-commit
    rev: v0.7.1
    hooks:
      - id: ruff
        name: Ruff lint
        args: [--fix, --exit-non-zero-on-fix]
      - id: ruff
        name: Ruff isort
        args: [check, --select, I, --fix]
      - id: ruff-format
        name: Ruff format
```

**Comment by thclark** [comment]
# New ruff settings for vscode devcontainers

I kept getting warnings about deprecated lsp settings for ruff. This is because they've added a new configuration setting. Totally needlessly as far as I can tell.

Your `devcontainer.json` should not contain any ruff settings, they should be read from pyproject.toml only.
```js
{
  // ...
  "customizations": {
    "vscode": {
      "settings": {
        "[python]": {
          "editor.defaultFormatter": "charliermarsh.ruff",
          // ...
        },
        "ruff.importStrategy": "fromEnvironment"
        // ...
      }
    }
  }
{
```
